addons/promote/common/src/main/java/org/commonjava/indy/promote/data/PromotionCallbackHelper.java

```diff
@@ -0,0 +1,144 @@
+
+import javax.annotation.PostConstruct;
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+
+import java.io.IOException;
+import java.util.Map;
+
+import static org.apache.http.client.utils.HttpClientUtils.closeQuietly;
+import static org.commonjava.indy.promote.model.CallbackTarget.CallbackMethod.POST;
+import static org.commonjava.indy.promote.model.CallbackTarget.CallbackMethod.PUT;
+import static org.commonjava.indy.util.ApplicationStatus.OK;
+
+/**
+ * Created by ruhan on 12/10/18.
+ */
+@ApplicationScoped
+public class PromotionCallbackHelper
+{
+    private final Logger logger = LoggerFactory.getLogger( getClass() );
+
+    private static final String PROMOTION_CALLBACK_SITE_ID = "promotion_callback"; // set it in http.conf
+
+    @Inject
+    private ObjectMapper objectMapper;
+
+    @Inject
+    private IndySiteConfigLookup siteConfigLookup;
+
+    private SiteConfig config;
+
+    private HttpFactory httpFactory;
+
+    public PromotionCallbackHelper()
+    {
+        this( new SiteConfigBuilder().withRequestTimeoutSeconds( 30 ).build(), new ObjectMapper() );
+    }
+
+    public PromotionCallbackHelper( SiteConfig config, ObjectMapper objectMapper )
+    {
+        this.config = config;
+        this.objectMapper = objectMapper;
+    }
+
+    @PostConstruct
+    private void init()
+    {
+        this.config = siteConfigLookup.lookup( PROMOTION_CALLBACK_SITE_ID );
+        ClientAuthenticator authenticator = null;
+        this.httpFactory = new HttpFactory( authenticator );
+    }
+
+    public <T extends AbstractPromoteResult> T callback( CallbackTarget target, T ret )
+    {
+        CloseableHttpClient client = null;
+        try
+        {
+            boolean isCallbackSuccess;
+            client = httpFactory.createClient( config );
+            if ( target.getMethod() == POST )
+            {
+                HttpPost req = new HttpPost( target.getUrl() );
+                isCallbackSuccess = execute( client, req, target, ret );
+            }
+            else if ( target.getMethod() == PUT )
+            {
+                HttpPut req = new HttpPut( target.getUrl() );
+                isCallbackSuccess = execute( client, req, target, ret );
+            }
+            else
+            {
+                throw new IllegalArgumentException( target.getMethod() + " not supported" );
+            }
+
+            if ( !isCallbackSuccess )
+            {
+                logger.warn( "Callback failed and stash it, target: {}", target );
+                stash( target, ret );
+            }
+        }
+        catch ( Exception e )
+        {
+            logger.error( "Callback failed", e );
+            stash( target, ret );
+        }
+        finally
+        {
+            closeQuietly( client );
+        }
+        return ret;
+    }
+
+    // TODO: add to queue and try later
+    private <T extends AbstractPromoteResult> void stash( CallbackTarget target, T ret )
```

Comment:
Is this intended to be implemented in the first pass?

Reply:
This is in progress. I studied some retry approaches (e.g., a home-made one, retry4j, infinispan) and would want to go for the infinispan one. I have a JIRA to track it. You may take a look at NOS-1586.

Reply:
I like the ISPN implementation and the backoff calculation. Looks great.